File: 0440-k-th-smallest-in-lexicographical-order/0440-k-th-smallest-in-lexicographical-order.go
package main

import (
	"fmt"
)

func findKthNumber(n int, k int) int {
	curr := 1
	k-- // Adjust k since we start counting from 1

	for k > 0 {
		count := countNumbers(n, curr)
		if count <= k {
			k -= count
			curr++ // Move to the next lexicographical number
		} else {
			k-- // Go deeper into the current "number" branch
			curr *= 10
		}
	}
	return curr
}

func countNumbers(n int, curr int) int {
	count := 0
	first := curr
	last := curr + 1

	for first <= n {
		count += min(n+1, last) - first
		first *= 10
		last *= 10
	}

	return count
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 0440-k-th-smallest-in-lexicographical-order/0440-k-th-smallest-in-lexicographical-order.java
class Solution {
    public int findKthNumber(int n, int k) {
        long current = 1;
        k--; // Convert to zero-based index
        
        while (k > 0) {
            long step = calculateSteps(current, current + 1, n);
            
            if (k >= step) {
                current++;
                k -= step;
            } else {
                current *= 10;
                k--;
            }
        }
        
        return (int) current;
    }
    
    private long calculateSteps(long curr, long next, int maxNum) {
        long steps = 0;
        while (curr <= maxNum) {
            steps += Math.min(next, maxNum + 1) - curr;
            curr *= 10;
            next *= 10;
        }
        return steps;
    }
}